probe-rs/src/flashing/loader.rs

```diff
@@ -429,6 +432,36 @@ impl FlashLoader {
         Ok(())
     }
 
+    /// Perform a Custom Chip-Erase if available and selected by the user.
+    ///
+    /// # Returns
+    /// Upon a successful Chip-Erase the download options will modifed to avoid any
+    /// further erase attempts.
+    pub(crate) fn do_custom_chip_erase_if_needed<'a>(
+        options: DownloadOptions<'a>,
+        session: &mut Session,
+    ) -> Result<DownloadOptions<'a>, FlashError> {
```

Comment:
I think placing this here and calling it in the loader butchers the division of responsibilities between structs a bit.

Reply:
Fair enough. This standalone function just takes care of calling the custom erase if applicable while updating progress and removing any further erase attempts later on.  Where would it fit better? It needs to be called early, before the Flasher is created since then the session gets turned into Core

Reply:
Hmm you should have access to the session in the flasher. Am I missing something here?
https://github.com/probe-rs/probe-rs/blob/master/probe-rs/src/flashing/loader.rs#L318 calls the regular erase on the flasher and it should be possible in a similar fashion. Tho I am not sure the flasher is the correct place to do it. TBH, the names and division of jobs is a bit crooked tbh.

Reply:
You were absolutely right. I somehow missed that the Flasher has access to the session. I moved it there and it becomes much cleaner IMHO